src/pkcs15init/pkcs15-lib.c

```diff
@@ -1294,6 +1345,11 @@ sc_pkcs15init_init_prkdf(struct sc_pkcs15_card *p15card, struct sc_profile *prof
 	r = SC_SUCCESS;
 
 err:
+	if (r < 0 && key->algorithm == SC_ALGORITHM_EC && key_info->params.data) {
+		free(((struct sc_ec_parameters *) key_info->params.data)->named_curve);
+		free(((struct sc_ec_parameters *) key_info->params.data)->der.value);
+		free(key_info->params.data);
+	}
```

Comment:
One suggestion for a simplification: Make `new_ecparams` a global variable set to `NULL` and keep the error handling here at the bottom of the function specific to `new_ecparams`. Above you don't need all the special cases anymore and you transfer ownership of the allocated memory to `key_info` if and only if the initialization was successful. If it was successful, set `new_ecparams` back to `NULL`.

Reply:
Thanks for pointing out, simplified.